qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,224 @@
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'y = y - x mod p'
+
+
+@frozen
+class ModNeg(Bloq):
+    r"""An n-bit modular negation gate.
+
+    Implements $U|x\rangle|p\rangle \rightarrow |-x mod p\rangle|p\rangle$ using $2n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('p', bitsize=self.bitsize)])
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, p: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'x = -x mod p'
+
+
+@frozen
+class ModDbl(Bloq):
+    r"""An n-bit modular negation gate.
+
+    Implements $U|x\rangle|p\rangle \rightarrow |2 * x mod p\rangle|p\rangle$ using $2n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('p', bitsize=self.bitsize)])
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2 * self.bitsize, Toffoli())}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, p: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+        return NotImplemented
+
+    def short_name(self) -> str:
+        return f'x = 2 * x mod p'
+
+
+@frozen
+class ModMult(Bloq):
+    r"""An n-bit modular multiplication gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle|0\rangle|0\rangle \rightarrow |x\rangle|y\rangle|p\rangle|garbage\rangle|x * y mod p\rangle$ using $2.25n^2 + 9n T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input register (register y above).
+        p: A bitsize-sized input register (register p above).
+        garbage: A bitsize-sized input register (register garbage above).
```

Comment:
is garbage used elsewhere in the algorithm? If it's just an ancilla I would think the decomposition would allocate / free this, or it should have a side?

Reply:
It is just an ancilla, I will remove it as a register and add it back in as an ancilla to the decomposition when I implement it.